include/igl/per_corner_normals.cpp

```diff
@@ -1,108 +1,168 @@
+// Copyright (C) 2021 Alec Jacobson <alecjacobson@gmail.com>
 // Copyright (C) 2013 Alec Jacobson <alecjacobson@gmail.com>
 //
 // This Source Code Form is subject to the terms of the Mozilla Public License
 // v. 2.0. If a copy of the MPL was not distributed with this file, You can
 // obtain one at http://mozilla.org/MPL/2.0/.
 #include "per_corner_normals.h"
-
 #include "vertex_triangle_adjacency.h"
 #include "per_face_normals.h"
 #include "PI.h"
+#include "parallel_for.h"
 #include "doublearea.h"
-
-template <typename DerivedV, typename DerivedF, typename DerivedCN>
-IGL_INLINE void igl::per_corner_normals(
-  const Eigen::MatrixBase<DerivedV>& V,
-  const Eigen::MatrixBase<DerivedF>& F,
-  const double corner_threshold,
-  Eigen::PlainObjectBase<DerivedCN> & CN)
-{
-  using namespace Eigen;
-  using namespace std;
-  Eigen::Matrix<typename DerivedV::Scalar,Eigen::Dynamic,3> FN;
-  per_face_normals(V,F,FN);
-  vector<vector<int> > VF,VFi;
-  vertex_triangle_adjacency(V.rows(),F,VF,VFi);
-  return per_corner_normals(V,F,FN,VF,corner_threshold,CN);
-}
+#include <Eigen/Geometry>
 
 template <
   typename DerivedV,
   typename DerivedF,
-  typename DerivedFN,
   typename DerivedCN>
 IGL_INLINE void igl::per_corner_normals(
-  const Eigen::MatrixBase<DerivedV>& V,
-  const Eigen::MatrixBase<DerivedF>& F,
-  const Eigen::MatrixBase<DerivedFN>& FN,
-  const double corner_threshold,
+  const Eigen::MatrixBase<DerivedV> & V,
+  const Eigen::MatrixBase<DerivedF> & F,
+  const typename DerivedV::Scalar corner_threshold_degrees,
   Eigen::PlainObjectBase<DerivedCN> & CN)
 {
-  using namespace Eigen;
-  using namespace std;
-  vector<vector<int> > VF,VFi;
-  vertex_triangle_adjacency(V.rows(),F,VF,VFi);
-  return per_corner_normals(V,F,FN,VF,corner_threshold,CN);
+  Eigen::Matrix<Eigen::Index,Eigen::Dynamic,1> VF,NK;
+  vertex_triangle_adjacency(F,V.rows(),VF,NK);
+  return per_corner_normals(V,F,corner_threshold_degrees,VF,NK,CN);
 }
 
 template <
   typename DerivedV,
   typename DerivedF,
-  typename DerivedFN,
-  typename IndexType,
+  typename DerivedVF,
+  typename DerivedNI,
   typename DerivedCN>
 IGL_INLINE void igl::per_corner_normals(
-  const Eigen::MatrixBase<DerivedV>& /*V*/,
-  const Eigen::MatrixBase<DerivedF>& F,
-  const Eigen::MatrixBase<DerivedFN>& FN,
-  const std::vector<std::vector<IndexType> >& VF,
-  const double corner_threshold,
+  const Eigen::MatrixBase<DerivedV> & V,
+  const Eigen::MatrixBase<DerivedF> & F,
+  const typename DerivedV::Scalar corner_threshold_degrees,
+  const Eigen::MatrixBase<DerivedVF> & VF,
+  const Eigen::MatrixBase<DerivedNI> & NI,
   Eigen::PlainObjectBase<DerivedCN> & CN)
 {
-  using namespace Eigen;
-  using namespace std;
+  typedef typename DerivedV::Scalar Scalar;
+  typedef Eigen::Index Index;
+  // unit normals
+  Eigen::Matrix<Scalar,Eigen::Dynamic,3,Eigen::RowMajor> FN(F.rows(),3);
+  // face areas
+  Eigen::Matrix<Scalar,Eigen::Dynamic,1> FA(F.rows());
+  //for(Index f = 0;f<F.rows();f++)
```

Comment:
Let's remove the commented code. We can turn parallel_for into sequential for loops using the env variable `IGL_NUM_THREADS`.